### ISSUE TITLE
Python assertion failed on shutdown

### ISSUE THREAD

**Issue by calebdoxsey** [grpc/grpc#8514]
We're seeing an assertion failure after we run our test suite:

```
E1026 14:30:30.842274624    9404 subchannel_index.c:70]      assertion failed: c != NULL

Program received signal SIGABRT, Aborted.
0x00007ffff68e0035 in raise () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) bt
#0  0x00007ffff68e0035 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff68e379b in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007fffd6ade226 in current_ctx.part.0 () at src/core/ext/client_config/subchannel_index.c:70
#3  0x00007fffd6b52a8c in sck_avl_destroy () at src/core/ext/client_config/subchannel_index.c:136
#4  0x00007fffd6b044bf in unref_node () from /home/vagrant/python/local/lib/python2.7/site-packages/grpc/_cython/cygrpc.so
#5  0x00007fffd6b504a9 in grpc_client_config_shutdown () from /home/vagrant/python/local/lib/python2.7/site-packages/grpc/_cython/cygrpc.so
#6  0x00007fffd6b0a695 in grpc_shutdown () from /home/vagrant/python/local/lib/python2.7/site-packages/grpc/_cython/cygrpc.so
#7  0x00000000004bc6dc in Py_Finalize ()
#8  0x00000000004bc928 in Py_Exit ()
#9  0x00000000004bca5c in ?? ()
#10 0x00000000004998bd in PyRun_SimpleFileExFlags ()
#11 0x000000000049a9cf in Py_Main ()
#12 0x00007ffff68cb7ed in __libc_start_main () from /lib/x86_64-linux-gnu/libc.so.6
#13 0x000000000041bb31 in _start ()
```

Seems very similar too: https://github.com/grpc/grpc/issues/7090

**Comment by nathanielmanistaatgoogle** [comment]
Thank you for the report! [How about some more detail](http://www.chiark.greenend.org.uk/~sgtatham/bugs.html)? With what hardware platform(s), operating system(s), Python version(s), and gRPC Python version(s) are you observing the defect? Is it consistently reproducible? Are you able to share code that demonstrates the defect? How minimal are you able to make the code that demonstrates the defect?

**Comment by calebdoxsey** [comment]
Unfortunately it's going to be very difficult to provide much more detail than this. It doesn't always happen, and when it does its only in the context of our gitlab test machine, and only after running 20 minutes of python tests... tests that, for obvious reasons, I can't share the source code for here. Suffice it to say, it's a python application with a dozens of dependencies, and a convoluted build process. 

I'm attempting to find a smaller example that might reproduce the issue. (though this is probably the kind of bug where finding a way to reproduce it consistently points directly to the solution)

We are using gRPC version 1.0.0. I tried upgrading to 1.0.1rc1, and that seems to have helped, but other engineers are telling me it still occasionally fails with the segfault. It's some sort of race condition on shutdown.

I realize this is not helpful; I put it up here in the hopes that it might be obvious what was wrong, or perhaps someone else had an easier reproduction. Feel free to close the issue if its not actionable in its current state, and I will open a new one if I can figure out something more concrete.

**Comment by nathanielmanistaatgoogle** [comment]
It's helpful that you've now shared with us _why_ you weren't able to include much detail in your initial report. Let's leave this here and open for now; perhaps you'll be able to contribute more or perhaps someone else seeing the problem will find it and add their information to it.

**Comment by nathanielmanistaatgoogle** [comment]
@dgquintas: I realize there's not a lot here, but could you offer some hand-wavy speculation about the kinds of things that would cause `E1026 14:30:30.842274624    9404 subchannel_index.c:70]      assertion failed: c != NULL`?

**Comment by dgquintas** [comment]
After spending some time looking at the subchannel_index.c code, I have some ideas that I need to run through some people. I'll report with what I find, but there's nothing obviously wrong going on here.

**Comment by dgquintas** [comment]
Please ping if I haven't updated by Monday.

**Comment by calebdoxsey** [comment]
Thanks for taking the time to look into this. I've been pushing really hard for my team to switch to gRPC for all of our services, and your quick response and dedication in the face of an admittedly terrible bug report is exemplary.

After running gRPC 1.0.1 for the last week or so, I have not seen this issue re-occur. It never happened in production, though it did happen in our build environment. Not sure what changed between 1.0.0 and 1.0.1, but this may not be a bug anymore.

**Comment by dgquintas** [comment]
Thank you for your kind words @calebdoxsey. I'll put the investigation on hold, as the issue may indeed have been resolved in the 1.0.1 release. I'm closing the issue. Needless to say, please reopen if the issue resurfaces. 

By the way, for future reference, you can get more logging by having the `GRPC_VERBOSITY` environment variable set to `DEBUG` prior to running your binary. There's also a collection of "tracers" for different part of the Core you may want to enable, even though it's a bit of a chicken-and-egg problem knowing which ones to enable. For maximum log spam action, you can set the `GRPC_TRACE` environment variable to `all`. See https://github.com/grpc/grpc/blob/1a62ec83b7af9c4437164d10fa23197c64542382/doc/environment_variables.md for more details.
